test/api/model/web_auth_test.dart

```diff
@@ -0,0 +1,97 @@
+import 'package:checks/checks.dart';
+import 'package:convert/convert.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:zulip/api/model/web_auth.dart';
+
+import '../../example_data.dart' as eg;
+
+void main() {
+  group('WebAuthPayload', () {
+    const otp = '186f6d085a5621ebaf1ccfc05033e8acba57dae03f061705ac1e58c402c30a31';
+    final encryptedApiKey = debugEncodeApiKey(eg.selfAccount.apiKey, otp);
+    final wellFormed = Uri.parse(
+      'zulip://login?otp_encrypted_api_key=$encryptedApiKey'
+        '&email=self%40example&user_id=1&realm=https%3A%2F%2Fchat.example%2F');
+
+    test('basic happy case', () {
+      final payload = WebAuthPayload.parse(wellFormed);
+      check(payload)
+        ..otpEncryptedApiKey.equals(encryptedApiKey)
+        ..email.equals('self@example')
+        ..userId.equals(1)
+        ..realm.equals(Uri.parse('https://chat.example/'));
+      check(payload.decodeApiKey(otp)).equals(eg.selfAccount.apiKey);
+    });
+
+    // TODO(server-5) remove
+    test('legacy: no userId', () {
+      final queryParams = {...wellFormed.queryParameters}..remove('user_id');
+      final payload = WebAuthPayload.parse(
+        wellFormed.replace(queryParameters: queryParams));
+      check(payload)
+        ..otpEncryptedApiKey.equals(encryptedApiKey)
+        ..email.equals('self@example')
+        ..userId.isNull()
+        ..realm.equals(Uri.parse('https://chat.example/'));
+      check(payload.decodeApiKey(otp)).equals(eg.selfAccount.apiKey);
+    });
+
+    test('parse fails when an expected field is missing', () {
+      final queryParams = {...wellFormed.queryParameters}..remove('email');
+      final input = wellFormed.replace(queryParameters: queryParams);
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('parse fails when otp_encrypted_api_key is wrong length', () {
+      final queryParams = {...wellFormed.queryParameters}
+        ..['otp_encrypted_api_key'] = 'asdf';
+      final input = wellFormed.replace(queryParameters: queryParams);
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('parse fails when host is not "login"', () {
+      final input = wellFormed.replace(host: 'foo');
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('parse fails when scheme is not "zulip"', () {
+      final input = wellFormed.replace(scheme: 'https');
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('decodeApiKey fails when otp is wrong length', () {
+      final payload = WebAuthPayload.parse(wellFormed);
+      check(() => payload.decodeApiKey('asdf')).throws<FormatException>();
+    });
+  });
+
+  group('generateOtp', () {
+    test('smoke, and check all 256 byte values are used', () {
```

Comment:
I did a quick test to confirm this test runs quickly, and it does. Running this test file with or without this test case included, 3 runs each, the difference in runtime is less than the noise between runs:
```
$ time flutter test test/api/model/web_auth_test.dart 
00:09 +8: All tests passed!                                                         
time: 10.660s wall (13.258s u, 0.864s s)

$ time flutter test test/api/model/web_auth_test.dart 
00:01 +8: All tests passed!                                                         
time: 2.808s wall (3.231s u, 0.613s s)

$ time flutter test test/api/model/web_auth_test.dart 
00:01 +8: All tests passed!                                                         
time: 2.724s wall (3.155s u, 0.596s s)

$ time flutter test test/api/model/web_auth_test.dart 
00:01 +8: All tests passed!                                                         
time: 2.702s wall (3.165s u, 0.581s s)

$ time flutter test test/api/model/web_auth_test.dart --name WebA
00:01 +7: All tests passed!                                                         
time: 2.764s wall (3.275s u, 0.580s s)

$ time flutter test test/api/model/web_auth_test.dart --name WebA
00:01 +7: All tests passed!                                                         
time: 2.759s wall (3.110s u, 0.642s s)

$ time flutter test test/api/model/web_auth_test.dart --name WebA
00:01 +7: All tests passed!                                                         
time: 2.710s wall (3.188s u, 0.568s s)
```

(I made one warmup run before the runs I counted, though I was surprised how big the warmup effect was.)

In particular I'm pretty sure the runtime is <50ms, which is fine.

Reply:
Thanks!